// project/events.js
var dal = require('dal/project');
var client = require('gh/client');
var config = require('rc')('prohub');

exports.test = function(stream, data, token) {
  if (data.page && data.page == 'project') 
    return true;
};

exports.handler = function(stream, data, token) {

  if (data.project) {

    data.project.token = token;
    dal.getProject(token, data.project, stream, function(err, data) {

      if (err) {
        return stream.write({ 
          page: 'project', 
          error: err.message 
        });
      }
 
      data.org = config.org;
      stream.write({
        page: 'project',
        project: data
      });
    });
  }

  if (data.remove) {
    var projectId = data.remove.id;
    var repo = data.remove.repo;

    return dal.removeRepo(projectId, repo, function(err) {
      if (err) {
        return stream.write({
          page: 'project',
          error: err.message
        });
      }

      stream.write({
        page: 'project',
        remove: repo
      });
    });
  }

  if (data.repo) {

    if (!data.repo || !data.repo.user || !data.repo.repo) {
      data.error = 'Incorrect arguments';
      return stream.write(data);
    }

    data.repo.token = token;

    client.getRepo(data.repo, function(err, repo) {
      if (err) {
        data.error = 'No such repo!';
        return stream.write(data);
      }

      data.repo.url = [
        'https://github.com',
        data.repo.user,
        data.repo.repo
      ].join('/');

      dal.addRepo(data.repo, function(err) {
        if (err) {
          data.error = err.message;
        }
        stream.write(data);
      });
    });
  }
};
