// gh/client.js
var GitHubApi = require('github-cache');
var ASSERT = require('assert').ok;

var github = new GitHubApi({
    // required
    version: '3.0.0',
    // optional
    debug: true,
    protocol: 'https',
    timeout: 5000
});

function auth(token) {
  github.authenticate({
    type: 'oauth',
    token: token.access_token
  });
}

function assertRequiredArgs(opts) {
  ASSERT(opts.token);
  ASSERT(opts.user);
  ASSERT(opts.repo);
  auth(opts.token);
}

exports.getRepo = function(opts, cb) {
  assertRequiredArgs(opts);
  github.repos.get(opts, cb); 
};

exports.getIssues = function(opts, cb) {
  assertRequiredArgs(opts);
  github.issues.repoIssues(opts, cb); 
};

exports.getMilestones = function(opts, cb) {
  assertRequiredArgs(opts);
  github.issues.getAllMilestones(opts, cb); 
};

exports.getContributors = function(opts, cb) {
  assertRequiredArgs(opts);
  github.repos.getStatsContributors(opts, cb); 
};

exports.getOrg = function(token, opts, cb) {
  ASSERT(token);
  ASSERT(opts.org);
  auth(token);
  github.orgs.get(opts, cb);
};

exports.getCurrentUser = function(token, opts, cb) {
  ASSERT(token);
  auth(token);
  github.user.get({}, cb);
};
