// home/client.js
var dom = require('dom-events');
var remove = require('dom-remove');
var classList = require('class-list');

var addBtn = document.querySelector('.add');
var closeBtn = document.querySelector('.modal-close');
var okBtn = document.querySelector('.modal-ok');
var projectInput = document.querySelector('.project-name');
var responseP = document.querySelector('.response p');
var addModal = document.querySelector('.modal');
var projectsList = document.querySelector('ul.projects-list');
var addModalCL = classList(addModal);
var unauthorizedP = document.querySelector('.unauthorized p');

function streamReader(stream) {
  stream.on('json', function(data) {

    if (data.add) {
      if (data.add.error) {
        responseP.innerHTML = data.add.error;
        return;
      }
      location.reload();
    }
    else if (data.remove) {
      remove(document.getElementById(data.remove));
    }

    if (data.error) {
      unauthorizedP.innerHTML = data.error;
    }
  });
}

function Reload(stream) {
  streamReader(stream)
}

function Load(stream) {

  dom.on(addBtn, 'click', function() {
    return addModalCL.add('modal-show');
  });

  dom.on(closeBtn, 'click', function() {
    return addModalCL.remove('modal-show');
  });

  dom.on(projectsList, 'click', function(event) {
    if (event.target.tagName == 'I') {
      var id = event.target.getAttribute('data-id');
      stream.write({
        page: 'home',
        remove: id
      });
    }
  });

  dom.on(okBtn, 'click', function() {

    if (projectInput.value == '') {
      responseP.innerHTML = 'Must have a name';
      return;
    }

    stream.write({
      page: 'home',
      add: projectInput.value
    });
  });

  streamReader(stream)
  module.exports = Reload;
};

module.exports = Load;
