// home/requests.js
var jade = require('jade');
var paramify = require('paramify');
var config = require('rc')('prohub');

var dal = require('dal/project');
var header = require('header');
var footer = require('footer');
var summarybar = require('summarybar');

//
// specify a path and compile a template function.
// let's assume that "index.jade" is the main template
// for dashboard.
//
var templateFile = __dirname + '/templates/index.jade';
var body = jade.compileFile(templateFile);

//
// use whatever to test the url and credentials, state, etc.
//
exports.test = function(req, res) {
  var match = paramify(req.url);

  if (match('/home') || match('/')) {
    return true;
  }
};

//
// if the test passes the handler will be fired.
//

exports.handler = function(req, res) {

  dal.listProjects(function(err, data) {
    if (err) {
      res.statusCode = 500;
      return res.end(err);
    }

    var summarybar_args = {
      projects: Object.keys(data).length || 0,
      issues: 0,
      milestones: 0,
      contributors: 0,
      staleness: 0
    };

    var body_args = {
      org: config.org || 'Unnamed',
      projects: data,
      summarybar: summarybar(summarybar_args)
    };
    
    res.write(header());
    res.write(body(body_args));
    res.end(footer());

  });
};
